ironfish/src/workerPool/roundrobinqueue.ts

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { Job } from './job'
+import { WorkerMessageType } from './tasks/workerMessage'
+
+// Simple queue that will iterate over each type of job sequentially so that each job type
+// will be guaranteed not to be backed up by one type queueing many jobs. This does not
+// solve the issue of long-running jobs, however. Luckily, this is not something we
+// have had to worry about yet.
+export class RoundRobinQueue {
+  private queueMap: Map<WorkerMessageType, Array<Job>>
+  private lastMapIndex = 0
+
+  // Returns the total length of all queues
+  get length(): number {
+    let length = 0
+
+    for (const queue of this.queueMap.values()) {
+      length += queue.length
+    }
+
+    return length
+  }
+
+  // Instantiates the queue with a map of each key with an empty array value
+  constructor() {
+    this.queueMap = new Map()
+
+    // Numerical enums return all keys, then all values as strings when
+    // trying to enumerate them, so we need to grab only the valid numerical values.
+    const enumKeys = Object.keys(WorkerMessageType)
+      .map((v) => Number(v))
+      .filter((v) => !isNaN(v))
+
+    for (const key of enumKeys) {
+      this.queueMap.set(key, [])
+    }
+  }
+
+  // Executes a function with every job across all queues
+  forEach(fn: (j: Job) => void): void {
+    for (const queue of this.queueMap.values()) {
+      for (const job of queue) {
+        fn(job)
+      }
+    }
+  }
+
+  // Add a job to that type's queue
+  enqueue(type: WorkerMessageType, job: Job): void {
+    const typeQueue = this.queueMap.get(type)
+
+    if (!typeQueue) {
+      return
```

Comment:
I think it'd be fine to throw if the queue doesn't exist -- I don't think there's a point where we'd be trying to queue up a job that doesn't have a mapping, so it'd probably indicate a bug in our code